File: src/pipeline/responseCache.ts
// responseCache.ts — backend response-level cache for /api/generate.
//
// orchestrator.generate() runs a live, minutes-long pipeline (headless Opus
// planner + Meshy preview->refine). The per-asset Meshy cache does NOT help a
// repeated /api/generate call: the planner emits a fresh meshyPrompt every run,
// so the Meshy key always misses. Memoizing the WHOLE response keyed by the
// request (prompt + amendRounds) fixes it — the first run is live, every
// identical request after is served from disk in well under a second.
//
// This stores the orchestrator's RAW output, whose ready glbUrls are LOCAL Meshy
// cache paths (PRE-bridge). The /assets bridge (publishSceneAssets) runs AFTER
// generate() returns and re-links + rewrites on every serve, so the two layers
// compose and this module stays out of the bridge's way.
//
// The on-disk envelope, sha256 key scheme, atomic write, and throw-on-corrupt vs
// null-on-expected-staleness read split all come from the shared diskCache helper;
// this module supplies only the response-specific bits (dir, version, and the
// glbUrl-on-disk validation below).

import { existsSync } from "node:fs";

import type { GenerateResponse } from "../api/contract";
import { normalizePrompt } from "../meshy/cache.mjs";
import { createDiskCache } from "./diskCache";

// Bumped when the envelope shape or the orchestrator output contract changes.
// Folded into the key (so old entries become unreachable) AND checked on read
// (so a stale-version file regenerates instead of deserializing a wrong shape).
const CACHE_VERSION = 1;

const cache = createDiskCache<GenerateResponse>({
  dirName: "responses",
  label: "response",
  valueKey: "response",
  version: CACHE_VERSION,
  validate: validateResponse,
});

// sha256 of [version, amendRounds, normalizedPrompt].join("::"), first 16 hex
// chars.
export function deriveResponseKey(prompt: string, amendRounds: number): string {
  return cache.deriveKey([String(amendRounds), normalizePrompt(prompt)]);
}

// Returns the cached response for `key`, or null on an EXPECTED miss: no file, a
// stale CACHE_VERSION, or a referenced local GLB that has been evicted. THROWS on
// a corrupt file (unparseable JSON, or a malformed envelope) — that is a bug, not
// a miss, and must surface loudly rather than silently regenerate.
export function readCachedResponse(key: string): GenerateResponse | null {
  return cache.read(key);
}

// Persists `response` under `key` inside an envelope carrying CACHE_VERSION and
// human-readable provenance (prompt/normalizedPrompt/amendRounds/savedAt).
export function writeCachedResponse(
  key: string,
  meta: { prompt: string; amendRounds: number },
  response: GenerateResponse,
): void {
  cache.write(
    key,
    { prompt: meta.prompt, normalizedPrompt: normalizePrompt(meta.prompt), amendRounds: meta.amendRounds },
    response,
  );
}

// Value validator for the disk cache. THROWS on a malformed response envelope (a
// bug, not a miss). Then performs asset-existence validation, mirroring the Meshy
// cache's validCandidatesOnDisk: a ready glbUrl that is a LOCAL path (not an
// http(s) URL, not an already bridged /assets/ URL) must still be on disk. If any
// referenced GLB was evicted, the cached response is unusable -> returns false so
// the caller regenerates.
function validateResponse(value: unknown, path: string): boolean {
  if (typeof value !== "object" || value === null || !Array.isArray((value as { passes?: unknown }).passes)) {
    throw new Error(`Corrupt response cache ${path}: "response" is not an object with a "passes" array`);
  }
  const response = value as GenerateResponse;
  for (const pass of response.passes) {
    for (const obj of pass.sceneState.objects) {
      const glbUrl = obj.glbUrl;
      if (typeof glbUrl !== "string" || glbUrl.length === 0) {
        continue;
      }
      if (glbUrl.startsWith("http://") || glbUrl.startsWith("https://") || glbUrl.startsWith("/assets/")) {
        continue;
      }
      if (!existsSync(glbUrl)) {
        return false; // a referenced local GLB was evicted — regenerate
      }
    }
  }
  return true;
}

File: backend/src/pipeline/response_cache.py
"""Response-level cache for Dream3D /api/generate results."""

from pathlib import Path
from typing import Any

from meshy.cache import normalize_prompt
from pipeline.disk_cache import create_disk_cache
from scene.schema import GenerateResponse

RESPONSE_CACHE_VERSION = 1


def _validate_response(value: Any, path: str) -> bool:
    if not isinstance(value, dict) or not isinstance(value.get("passes"), list):
        raise RuntimeError(
            f'Corrupt response cache {path}: "response" is not an object with a "passes" array'
        )

    for pass_ in value["passes"]:
        for obj in pass_["sceneState"]["objects"]:
            glb_url = obj.get("glbUrl")
            if not isinstance(glb_url, str) or len(glb_url) == 0:
                continue
            if (
                glb_url.startswith("http://")
                or glb_url.startswith("https://")
                or glb_url.startswith("/assets/")
            ):
                continue
            if not Path(glb_url).exists():
                return False
    return True


cache = create_disk_cache(
    dir_name="responses",
    label="response",
    value_key="response",
    version=RESPONSE_CACHE_VERSION,
    validate=_validate_response,
)


def derive_response_key(prompt: str, amend_rounds: int) -> str:
    return cache.derive_key([str(amend_rounds), normalize_prompt(prompt)])


def read_cached_response(key: str) -> GenerateResponse | None:
    value = cache.read(key)
    if value is None:
        return None
    return GenerateResponse.model_validate(value)


def write_cached_response(
    key: str,
    *,
    prompt: str,
    amend_rounds: int,
    response: GenerateResponse,
) -> None:
    cache.write(
        key,
        {
            "prompt": prompt,
            "normalizedPrompt": normalize_prompt(prompt),
            "amendRounds": amend_rounds,
        },
        response.model_dump(by_alias=True),
    )
